shell/platform/linux/fl_engine.cc

```diff
@@ -0,0 +1,163 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/linux/public/flutter_linux/fl_engine.h"
+#include "flutter/shell/platform/linux/fl_engine_private.h"
+#include "flutter/shell/platform/linux/fl_renderer.h"
+
+#include <gmodule.h>
+
+struct _FlEngine {
+  GObject parent_instance;
+
+  FlDartProject *project;
+  FlRenderer* renderer;
+  FLUTTER_API_SYMBOL(FlutterEngine) engine;
+};
+
+G_DEFINE_TYPE(FlEngine, fl_engine, G_TYPE_OBJECT)
+
+/* Callback from Flutter engine that are passed to the renderer */
+
+static void* fl_engine_gl_proc_resolver(void* user_data, const char* name) {
+  FlEngine* self = static_cast<FlEngine*>(user_data);
+  return fl_renderer_get_proc_address(self->renderer, name);
+}
+
+static bool fl_engine_gl_make_current(void* user_data) {
+  FlEngine* self = static_cast<FlEngine*>(user_data);
+  g_autoptr(GError) error = NULL;
+  gboolean result = fl_renderer_make_current(self->renderer, &error);
+  if (!result)
+    g_warning ("%s", error->message);
+  return result;
+}
+
+static bool fl_engine_gl_clear_current(void* user_data) {
+  FlEngine* self = static_cast<FlEngine*>(user_data);
+  g_autoptr(GError) error = NULL;
+  gboolean result = fl_renderer_clear_current(self->renderer, &error);
+  if (!result)
+    g_warning ("%s", error->message);
+  return result;
+}
+
+static uint32_t fl_engine_gl_fbo_callback(void* user_data) {
+  FlEngine* self = static_cast<FlEngine*>(user_data);
+  return fl_renderer_get_fbo(self->renderer);
+}
+
+static bool fl_engine_gl_present(void* user_data) {
+  FlEngine* self = static_cast<FlEngine*>(user_data);
+  g_autoptr(GError) error = NULL;
+  gboolean result = fl_renderer_present(self->renderer, &error);
+  if (!result)
+    g_warning ("%s", error->message);
+  return result;
+}
+
+static void fl_engine_dispose(GObject* object) {
+  FlEngine* self = FL_ENGINE(object);
+
+  g_clear_object(&self->project);
+  g_clear_object(&self->renderer);
+
+  FlutterEngineDeinitialize(self->engine);
```

Comment:
The first thing `FlutterEngineShutdown` does is call this, and that's documented behavior, so you can remove this call.